server/src/main/java/io/crate/auth/JWTAuthenticationMethod.java

```diff
@@ -0,0 +1,174 @@
+            verifier.verify(decodedJWT);
+
+        } catch (Exception e) {
+            throw new RuntimeException(
+                String.format(
+                    Locale.ENGLISH,
+                    "jwt authentication failed for user %s. Reason: %s",
+                    username,
+                    e.getMessage()
+                )
+            );
+        }
+
+        return user;
+    }
+
+
+    public static JwkProvider jwkProvider(@NotNull String issuer) {
+        URL jwkUrl;
+        try {
+            /*
+             We cannot use JwkProviderBuilder constructor with String.
+             It adds hard coded WELL_KNOWN_JWKS_PATH (/.well-known/jwks.json) to the url.
+             JWK URL not necessarily ends with that suffix, for example:
+             Microsoft: "jwks_uri":" https://login.microsoftonline.com/common/discovery/v2.0/keys"
+             Google: "jwks_uri": "https://www.googleapis.com/oauth2/v3/certs",
+             Taken from https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration
+             and https://accounts.google.com/.well-known/openid-configuration correspondingly
+            */
+
+            URI uri = new URI(issuer).normalize();
+            jwkUrl = uri.toURL();
+        } catch (MalformedURLException | URISyntaxException e) {
+            throw new IllegalArgumentException("Invalid jwks uri", e);
+        }
+
+        return new JwkProviderBuilder(jwkUrl).build();
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    /**
+     * Resolved algorithm based on the endpoint info.
+     * Cannot resolve algorithm based on the token header as it's optional and unsafe.
+     * If token has "alg" claim, it's checked against JWK info and in case of mismatch, error is thrown.
+     * If JWK info doesn't have "alg" (it's optional for endpoint response as well), we fall back to RSA256.
+     */
+    private static Algorithm resolveAlgorithm(JwkProvider jwkProvider,
+                                              @NotNull DecodedJWT decodedJWT) throws JwkException {
+        Jwk jwk = jwkProvider.get(decodedJWT.getKeyId());
+        PublicKey publicKey = jwk.getPublicKey();
+        if (publicKey instanceof RSAPublicKey == false) {
+            throw new UnsupportedOperationException("Only RSA algorithm is supported for JWT");
+        }
+        RSAPublicKey rsaPublicKey = (RSAPublicKey) publicKey;
+
+        // Default.
+        if (jwk.getAlgorithm() == null) {
+            return Algorithm.RSA256(new LoadedRSAKeyProvider(rsaPublicKey));
+        }
+
+        // Try to validate resolved info with token (if both JWK and token are not null).
+        if (decodedJWT.getAlgorithm() != null && decodedJWT.equals(jwk.getAlgorithm()) == false) {
+            throw new IllegalArgumentException("Jwt token has algorithm not matching with the algorithm of the public key.");
+        }
+
+        return switch (jwk.getAlgorithm()) {
```

Comment:
shouldn't we do uppercase before checking the values?

Reply:
This is alg from the JWK endpoint. By spec it's case sensitive

https://datatracker.ietf.org/doc/html/rfc7517#section-4.4

>The "alg" value is a case-sensitive ASCII string.

If not specified (like in https://console.cratedb-dev.cloud/api/v2/meta/jwk/ or in https://login.microsoftonline.com/common/discovery/v2.0/keys) we go with the default RSA256

Otherwise it should be uppercase - I checked https://www.googleapis.com/oauth2/v3/certs and also jwt library Algorithm.RSA... ctors

Reply:
great, thx a lot!